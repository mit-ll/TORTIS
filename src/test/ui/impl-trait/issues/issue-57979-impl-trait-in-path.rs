// rust-lang/rust#57979 : the initial support for `impl Trait` didn't
// properly check syntax hidden behind an associated type projection.
// Here we test behavior of occurrences of `impl Trait` within a path
// component in that context.

mod allowed {
    #![allow(nested_impl_trait)]

    pub trait Bar { }
    pub trait Quux<T> { type Assoc; }
    pub fn demo(_: impl Quux<(), Assoc=<() as Quux<impl Bar>>::Assoc>) { }
    impl<T> Quux<T> for () { type Assoc = u32; }
}

mod warned {
    #![warn(nested_impl_trait)]

    pub trait Bar { }
    pub trait Quux<T> { type Assoc; }
    pub fn demo(_: impl Quux<(), Assoc=<() as Quux<impl Bar>>::Assoc>) { }
    //~^ WARN `impl Trait` is not allowed in path parameters
    //~| WARN will become a hard error in a future release!
    impl<T> Quux<T> for () { type Assoc = u32; }
}

mod denied {
    #![deny(nested_impl_trait)]

    pub trait Bar { }
    pub trait Quux<T> { type Assoc; }
    pub fn demo(_: impl Quux<(), Assoc=<() as Quux<impl Bar>>::Assoc>) { }
    //~^ ERROR `impl Trait` is not allowed in path parameters
    //~| WARN will become a hard error in a future release!
    impl<T> Quux<T> for () { type Assoc = u32; }
}

fn main() { }
