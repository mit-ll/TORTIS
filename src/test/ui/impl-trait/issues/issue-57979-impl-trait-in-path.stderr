warning: `impl Trait` is not allowed in path parameters
  --> $DIR/issue-57979-impl-trait-in-path.rs:20:52
   |
LL |     pub fn demo(_: impl Quux<(), Assoc=<() as Quux<impl Bar>>::Assoc>) { }
   |                                                    ^^^^^^^^
   |
note: lint level defined here
  --> $DIR/issue-57979-impl-trait-in-path.rs:16:13
   |
LL |     #![warn(nested_impl_trait)]
   |             ^^^^^^^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #59014 <https://github.com/rust-lang/rust/issues/59014>

error: `impl Trait` is not allowed in path parameters
  --> $DIR/issue-57979-impl-trait-in-path.rs:31:52
   |
LL |     pub fn demo(_: impl Quux<(), Assoc=<() as Quux<impl Bar>>::Assoc>) { }
   |                                                    ^^^^^^^^
   |
note: lint level defined here
  --> $DIR/issue-57979-impl-trait-in-path.rs:27:13
   |
LL |     #![deny(nested_impl_trait)]
   |             ^^^^^^^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #59014 <https://github.com/rust-lang/rust/issues/59014>

error: aborting due to previous error

